rubysg/rubysg-reboot
# Fix broken RubyWeekly message

# Description 

The weekly RubyWeekly message by the RubySgBot has been broken for many weeks. The purpose of this PR is to fix the broken formatting, and to bring back a readable weekly message.

## What it's like now.
<img width="500" alt="Screenshot 2020-12-18 at 12 10 20 AM" src="https://user-images.githubusercontent.com/10722197/102512831-71634c00-40c5-11eb-9f85-f1e2cfdc5a58.png">

## What it's like after this PR
<img width="500" alt="Screenshot 2020-12-18 at 12 13 16 AM" src="https://user-images.githubusercontent.com/10722197/102513176-d9199700-40c5-11eb-8b8e-eec260eea0b0.png">


# What caused the markdown formatting to break?

The markdown formatting is broken whenever there is a hash character (`#`) in the message. It is broken for some unknown reason that I couldn't find out. 


# Solution 

Switched out custom `TelegramBotApi` with `telegram-bot` gem. After reaching out to some folks at WIP Ruby, they recommended to use the `telegram-bot` gem and somehow the message with the hash character (`#`) is parsed properly. 


# Improvements on the side

* Split 5 article summaries into 1 telegram message each. 
    * While investigating this problem, I also chanced upon another problem about max number of characters per message sent. For this week's RubyWeekly issue, the article is exceptionally long and the resulting message is over the 4096 character limit by Telegram's API. 
    * Even if the current implementation still works, the api would run into the character limit error.
    * This implementation would result in many messages sent and might become "spam", however I would like to leave the optimizing work for another PR given this PR is pretty big already. 

# How to test? 

1. Set up Ngrok and open a Ngrok process to your local rubysg-reboot server. This can be done by sending a GET request to `https://api.telegram.org/<bot-token>/setWebhook?url=https://<sub-domain>.ngrok.io/api/ruby_sg_bot/webhook`. You can get the bot token from me personally.
2. Create a telegram group and add RubySGBot to your channel. You should see a new entry in `RubySgBotSubscriber` table in your DB. 
3. Open a console and run the following code. RubySGBot should now fetch the latest RubyWeekly content, parse it and send it to your telegram group. 
    ```
    SendWeeklyUpdatesJob.new.perform_now(chat_id: RubySgBotSubscriber.last.chat_id)
    ```
4. Set RubySGBot back to the original webhook by sending a GET request to `https://api.telegram.org/<bot-token>/setWebhook?url=https://ruby.sg/api/ruby_sg_bot/webhook`
